Unpin the `click` dependency

sphuber opened the pull request:
Fixes #3838 
Fixes #3810 

The breaking of the tests was caused by two separate commits:

  * 718485be48263056e7036ea9a60ce11b47e2fc26
  * 37d897069f58f7f2a016c14b0620c6d387430b4b

The first one changes the format of the output of a command if a
required option is missing. The double quotes around the option have
been changed to single quotes.

The second commit is more pernicious: the `Editor.edit_file` method was
updated to escape the editor and filename parameters passed to the sub
process call. Our tests were abusing the absence of escaping to pass
arguments to the editor command, in our case `vim`, as to make the
normally interactive command a non-interactive one for testing purposes.
Now that the editor command is escaped, the arguments are understood by
bash to be part of the command which of course then cannot be found and
the test fails. We "fix" this problem by patching the changed method in
`click` to undo the escaping.

sphuber commented:
/update-requirements

sphuber commented:
@ltalirz @greschd and @csadorf note that in the first commit I fix the version of `click` to be equal higher `7.1`. This is just temporarily to force the tests to run on that version, since they were broken. After review, when you think this is good to be merged, I can revert the requirement to `click~=7.0` since it should now be compatible with both minor versions.

ltalirz commented on tests/conftest.py:
```diff
@@ -11,3 +11,48 @@
+def non_interactive_editor(request):
+    """Fixture to patch click's `Editor.edit_file`.
+
+    In `click==7.1` the `Editor.edit_file` command was changed to escape the `os.environ['EDITOR']` command. This was
+    breaking our tests since were abusing this system to define an automated vim command in order to make an interactive
+    command non-interactive. However, the escaping made the command and arguments understood as a single command and
+    so bash could not find it. Here we patch the method to remove the escaping of the editor command.
```
```suggestion
    In `click==7.1` the `Editor.edit_file` command was changed to escape the `os.environ['EDITOR']` command.
    We are currently abusing this variable to define an automated vim command in order to make an interactive
    command non-interactive, and escaping it makes bash interpret the command and its arguments as a single command instead.
    Here we patch the method to remove the escaping of the editor command.
```

ltalirz replied:
Why is it enough to define this as a fixture? Is this relevant just for testing?
Perhaps I haven't fully grasped the problem here...

sphuber replied:
I am pretty sure yeah, because I don't think we explicitly pass arguments to a file editor anywhere else in the code. But cannot be 100% sure

ltalirz replied:
Ok. Just checking whether you are still planning to accept the changes, which are meant describe the status quo rather than what changed.

ltalirz commented on tests/cmdline/commands/test_code.py:
```diff
@@ -276,3 +231,54 @@ def setUp(self):
         self.assertEqual(1, result.output.count('# No codes found matching the specified criteria.'))
+
+
+@pytest.mark.parametrize('non_interactive_editor', ('sleep 1; vim -cwq',), indirect=True)
```
ok, I guess it has to do with this monkey-patching fixture...

sphuber replied:
Yes, this is the way to paramterize a pytest fixture. In this case it parameterizes the command of the `non_interactive_editor` fixture

sphuber commented on tests/cmdline/params/options/test_conditional.py:
```diff
@@ -66,7 +66,10 @@ def test_switch_on(self):
         result = runner.invoke(cmd, ['--on'])
         self.assertIsNotNone(result.exception)
-        self.assertIn('Error: Missing option "--opt".', result.output)
+        # The quoting changed in `click==7.1` so we have to check for both cases if we want to support `click<7.1`
```
in any case, the condition would require two clauses, but the alternative would maybe indeed be shorter and more legible. Something like:
```
assert "Error: Missing option" in result.output and "--opt" in result.output
```

sphuber commented on tests/cmdline/commands/test_code.py:
```diff
@@ -276,3 +231,54 @@ def setUp(self):
+    assert isinstance(Code.get_from_string('{}@{}'.format(label, aiida_localhost.name)), Code)
+
+
+@pytest.mark.skip('This test is hanging for some reason.')
```
I tried fixing it but I had even trouble understanding what the problem was. The problem was that the test hangs and you get no output. By changing the `click` source and setting default of `resilient_parsing=False` for the `Context` at least it printed a warning. Apparently the `CodeBuilder` was flagging an issue saying that properties were specified for a "remote" code even though it thinks that it concerns a local one. This is weird, because in the test what we are duplicating is a "remote" code. Anyway, I find the code of the `CodeBuilder` a bit hard to follow and I could not figure out why now all of a sudden it was complaining when I am just switching to a different manner of running the test and the test itself remaining essentially unchanged. Would be great if you may have some pointers, since I think you are the original author of the `CodeBuilder`.

ltalirz replied:
> since I think you are the original author of the CodeBuilder

I think this honor belongs to Rico, in 6873af9 .
However perhaps the issue is indeed in the code I added for code duplication?
E.g. the `Code.get_code_spec` function?
https://github.com/aiidateam/aiida-core/blob/0644e7d9ff492fdb2c94d608c575c697daee6d32/aiida/orm/utils/builders/code.py#L88-L115

Anyhow, since Codes are now Nodes, perhaps this function is basically obsolete and one could simply clone the Code?=

ltalirz replied:
Replacing this with a `clone` is probably still a valid point, perhaps open an issue for this if you agree?

sphuber replied:
I was going to, but it does quite a bit more then just providing the functionality that `clone` does. It also provides all the validation and a way for the contextual defaults to be set in the CLI, so I am not sure that we can actually get rid of it.

sphuber commented:
@ltalirz I fixed the hanging test and in doing so tackled another issue #3810 . If you are happy with the changes let me know, then I will rebase and revert the requirements to `click~=7.0`. It would maybe be good to also merge your PR #3894 before this, as it touches the test that was hanging and I now fixed.

ltalirz commented on aiida/manage/tests/pytest_fixtures.py:
```diff
@@ -118,55 +115,75 @@ def test_1(aiida_localhost):
+        label=None,
+        prepend_text=None,
+        append_text=None,
+        check_executable=True
```
My only comment is that we should be a bit careful not to let the parameters of this function explode.
Is it really necessary to use this fixture to set up a code pointing to a non-existing path?

sphuber replied:
I would say yes, in most test case scenarios, the executable is never actually run so who cares if it can actually be found. So to be honest I am not sure why the check is there, but in order to not break backwards compatibility I put in a flag.

sphuber replied:
I have reverted the change for now

ltalirz commented on tests/cmdline/commands/test_code.py:
```diff
@@ -276,3 +231,52 @@ def setUp(self):
+    """Test code duplication interactive."""
+    label = 'code_duplicate_interactive_teor'
+    user_input = label + '\n\n\n\n\n\n'
+    code = aiida_local_code_factory('arithmetic.add', '/remote/abs/path', label='code', check_executable=False)
```
e.g. here, you could point to an existing executable, say `python` or `verdi` or ...
